compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,139 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms.{ MiniPhaseTransform, TransformerInfo }
+import ast.Trees._, ast.tpd, core._
+import Contexts.Context, Types._, Decorators._, Symbols._, DenotTransformers._
+import SymDenotations._, Scopes._, StdNames._, NameOps._, Names._
+
+import scala.collection.mutable
+
+/** Specializes classes that inherit from `FunctionN` where there exists a
+ *  specialized form.
+ */
+class SpecializeFunctions extends MiniPhaseTransform with InfoTransformer {
+  import ast.tpd._
+  val phaseName = "specializeFunctions"
+  override def runsAfter = Set(classOf[ElimByName])
+
+  /** Transforms the type to include decls for specialized applys  */
+  def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context) = tp match {
+    case tp: ClassInfo if !sym.is(Flags.Package) && (tp.decls ne EmptyScope) =>
+      var newApplys = Map.empty[Name, Symbol]
+
+      tp.parents.foreach { parent =>
+        List(0, 1, 2).foreach { arity =>
+          val func = defn.FunctionClass(arity)
+          if (parent.derivesFrom(func)) {
+            val typeParams = tp.cls.typeRef.baseType(func).argInfos
+            val interface  = specInterface(typeParams)
+
+            if (interface.exists && tp.decls.lookup(nme.apply).exists) {
+              val specializedMethodName = nme.apply.specializedFunction(typeParams.last, typeParams.init)
+              newApplys += (specializedMethodName -> interface)
+            }
+          }
+        }
+      }
+
+      def newDecls =
+        newApplys.toList.map { case (name, interface) =>
+          ctx.newSymbol(
+            sym,
+            name,
+            Flags.Override | Flags.Method | Flags.Synthetic,
+            interface.info.decls.lookup(name).info
+          )
+        }
+        .foldLeft(tp.decls.cloneScope) {
+          (scope, sym) => scope.enter(sym); scope
+        }
+
+      if (newApplys.isEmpty) tp
+      else tp.derivedClassInfo(decls = newDecls)
+
+    case _ => tp
+  }
+
+  /** Transforms the `Template` of the classes to contain forwarders from the
+   *  generic applys to the specialized ones. Also inserts the specialized applys
+   *  in the template body.
+   */
+  override def transformTemplate(tree: Template)(implicit ctx: Context, info: TransformerInfo) = {
+    val applyBuf = new mutable.ListBuffer[Tree]
+    val newBody = tree.body.mapConserve {
+      case dt: DefDef if dt.name == nme.apply && dt.vparamss.length == 1 =>
+        val specName = nme.apply.specializedFunction(
+          dt.tpe.widen.finalResultType,
+          dt.vparamss.head.map(_.symbol.info)
+        )
+
+        val specializedApply = tree.symbol.enclosingClass.info.decls.lookup(specName)
```

Comment:
Same thing here and in transformApply, you should only have to do the lookup if `isSpecializableFunction` returns true